Allow users to escape ".", "[", or "]" in the meta primary

ekinanp opened the pull request:
Fixes #509

ekinanp committed:
Move key sequence parsing to parseKeySequence

This will make it easier to support the subsequent work on escaping
'.', '[', ']'

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Allow users to escape ".", "[", or "]" in the meta primary

Fixes https://github.com/puppetlabs/wash/issues/509

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
```
wash . ❯ find docker -k '*container' -meta '.labels.com\.docker\.compose\.config-hash' -exists
docker/containers/wash_tutorial_redis_1
docker/containers/wash_tutorial_web_1
```

MikaelSmith commented on cmd/internal/find/primary/meta.go:
```diff
@@ -172,6 +172,9 @@ matches 'foo'.
       p(o['bar']) returns true.
 
+NOTE: You can use a backslash "\" to escape ".", "[", or "]". For example,
+'.foo\.bar p' returns p(m['foo.bar']).
```
Where does the `m` come from in this key sequence?

ekinanp replied:
The examples above the `NOTE` use `m` to reference the entry's metadata, so I think people shouldn't have trouble figuring that out.

MikaelSmith commented on cmd/internal/find/primary/meta/objectPredicate.go:
```diff
@@ -71,6 +53,51 @@ func parseObjectP(tokens []string, baseCaseParser, keySequenceParser predicate.P
+// NOTE: Users can still specify ".", "[", or "]" by escaping
+// them with a backslash "\". For example, '.com\.docker\.compose'
+// would be parsed as the key "com.docker.compose".
+func parseKeySequence(tk string) (string, string, error) {
```
The name could reflect that it's only parsing part of the string, like `nextKeyFromSequence`.

ekinanp replied:
Renamed to `parseKey` -- I think the return type should make it clear that it's parsing part of the string to get the key.

MikaelSmith commented on cmd/internal/find/primary/meta/objectPredicate.go:
```diff
@@ -71,6 +53,51 @@ func parseObjectP(tokens []string, baseCaseParser, keySequenceParser predicate.P
+				rem = rem[2:]
+				continue
+			}
+			// Otherwise, include "\" as part of the key
```
This means you can't have a sequence like `.slashy\.mcslash` where the backslash is part of the first key. That's probably fine.

ekinanp replied:
Thanks, updated so that "\"'es are also escapable.

MikaelSmith commented on cmd/internal/find/primary/meta/objectPredicate.go:
```diff
@@ -71,6 +53,51 @@ func parseObjectP(tokens []string, baseCaseParser, keySequenceParser predicate.P
+		if rem[0] == '\\' {
+			// Lookahead and check for an escaped character
+			if len(rem) > 1 && isTerminatingChar(rem[1]) {
+				key += string(rem[1])
```
These 3 lines could be `rem = rem[1:]` and still work. It says: this is a valid escape, remove the backslash and continue to add it to the key.

ekinanp replied:
Updated.

ekinanp committed:
Change parseKeySequence => parseKey

Latter makes what we're parsing clearer.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Also escape "\"'es in meta primary

This lets people type keys like "foo\". We also simplify the parsing
semantics by erroring if a "\" isn't followed by an escapable character.

Signed-off-by: Enis Inan <enis.inan@puppet.com>
